tests/unit/panther_analysis_tool/test_testing.py

```diff
@@ -0,0 +1,304 @@
+it under the terms of the GNU Affero General Public License as
+published by the Free Software Foundation, either version 3 of the
+License, or (at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+
+
+import unittest
+
+from panther_analysis_tool.detection import DetectionResult
+from panther_analysis_tool.rule import Rule, TYPE_RULE
+
+from panther_analysis_tool.testing import FunctionTestResult, TestError, TestSpecification, \
+    TestExpectations, TestCaseEvaluator, TestResult, TestResultsPerFunction
+
+TEST_RULE = {
+    'body': 'def rule(_):\n\treturn True',
+    'id': 'test-id',
+    'severity': 'INFO',
+    'type': TYPE_RULE,
+    'versionId': 'my-version',
+}
+
+class TestFunctionTestResult(unittest.TestCase):
+
+    def test_new(self) -> None:
+        # If output is boolean
+        result = FunctionTestResult.new(output=True)
+        self.assertEqual(result, FunctionTestResult(output='true', error=None))
+
+        # If output is string
+        result = FunctionTestResult.new(output='some output')
+        self.assertEqual(result, FunctionTestResult(output='some output', error=None))
+
+        # If both parameters are None
+        result = FunctionTestResult.new(output=None, raw_exception=None)
+        self.assertIsNone(result)
+
+        # When an exception is given
+        exception = TypeError('wrong type')
+        result = FunctionTestResult.new(output='some output', raw_exception=exception)
+        expected = FunctionTestResult(output='some output', error=TestError(message='TypeError: wrong type'))
+        self.assertEqual(result, expected)
+
+    def test_format_exception(self) -> None:
+        self.assertIsNone(FunctionTestResult.format_exception(exc=None))
+
+        # If title is None
+        exception = TypeError('wrong type')
+        self.assertEqual(FunctionTestResult.format_exception(exc=exception), 'TypeError: wrong type')
+        self.assertEqual(FunctionTestResult.format_exception(exc=exception, title='invalid type'), 'invalid type: TypeError: wrong type')
+
+    def test_to_test_error(self) -> None:
+        self.assertIsNone(FunctionTestResult.to_test_error(exc=None))
+        exception = TypeError('wrong type')
+        self.assertEqual(FunctionTestResult.to_test_error(exc=exception), TestError(message='TypeError: wrong type'))
+        self.assertEqual(
+            FunctionTestResult.to_test_error(exc=exception, **dict(title='invalid type')),
+            TestError(message='invalid type: TypeError: wrong type')
+        )
+
+
+class TestTestCaseEvaluator(unittest.TestCase):
+
+
```

Comment:
`[minor]` notice you have two spaces here ... but one above, not sure which is right

Reply:
fixed